python/lsst/daf/butler/_dataset_ref.py

```diff
@@ -1019,14 +1046,27 @@ def to_refs(self, universe: DimensionUniverse) -> list[DatasetRef]:
             name: DatasetType.from_simple(dtype, universe=universe)
             for name, dtype in self.dataset_types.items()
         }
+
+        # Dimension records can be attached if available.
+        # We assume that all dimension information was stored.
+        attacher = None
+        if self.dimension_records:
+            attacher = DimensionDataAttacher(
+                deserializers=self.dimension_records.make_deserializers(universe)
+            )
+
         refs: list[DatasetRef] = []
         for id_, minimal in self.compact_refs.items():
+            dataset_type = dataset_types[minimal.dataset_type_name]
             simple_data_id = SerializedDataCoordinate(dataId=minimal.data_id)
             data_id = DataCoordinate.from_simple(simple=simple_data_id, universe=universe)
+            if attacher:
+                data_ids = attacher.attach(dataset_type.dimensions, [data_id])
```

Comment:
At first I was a little worried this would be less efficient than calling it with a lot of data IDs at once, but I think that's only true if you pass a `Query` in.  So for now just something to keep in mind; I don't think it's worth it to try to work this code to be able to pass in multiple data IDs at once.